[ubsan] Display correct runtime messages for negative _BitInt

earnol opened the pull request:
Without this patch compiler-rt ubsan library has a bug displaying incorrect values for variables of the _BitInt (previously called _ExtInt) type. This patch affects affects both: generation of metadata inside code generator and runtime part. The runtime part provided only for i386 and x86_64 runtimes. Other runtimes should be updated to take full benefit of this patch.
The patch is constructed the way to be backward compatible and int and float type runtime diagnostics should be unaffected for not yet updated runtimes.

This patch fixes issue https://github.com/llvm/llvm-project/issues/64100.

MaskRay commented on compiler-rt/test/ubsan/TestCases/Integer/bit-int.c:
```diff
@@ -0,0 +1,174 @@
+
+uint32_t integer_divide_by_zero() __attribute__((no_sanitize("memory"))) {
+  _BitInt(37) x = 1 / 0;
+  // RUNTIME: {{.*}}bit-int.c:[[@LINE-1]]:21: runtime error: division by zero
```
`[[@LINE]]` is legacy. Prefer `[[#@LINE]]` for new code https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-pseudo-numeric-variables

earnol replied:
Accepted.

MaskRay commented on compiler-rt/test/ubsan/TestCases/Integer/bit-int.c:
```diff
@@ -0,0 +1,174 @@
+// RUN: %clang -Wno-constant-conversion -Wno-array-bounds -Wno-division-by-zero -Wno-shift-negative-value -Wno-shift-count-negative -Wno-int-to-pointer-cast -O0 -fsanitize=alignment,array-bounds,bool,float-cast-overflow,implicit-integer-sign-change,implicit-signed-integer-truncation,implicit-unsigned-integer-truncation,integer-divide-by-zero,nonnull-attribute,null,nullability-arg,nullability-assi ...
+// RUN: %clang -Wno-constant-conversion -Wno-array-bounds -Wno-division-by-zero -Wno-shift-negative-value -Wno-shift-count-negative -Wno-int-to-pointer-cast -fsanitize=array-bounds,enum,float-cast-overflow,integer-divide-by-zero,implicit-unsigned-integer-truncation,implicit-signed-integer-truncation,implicit-integer-sign-change,unsigned-integer-overflow,signed-integer-overflow,shift-base,shift-ex ...
```
compiler-rt/test is for runtime tests. IR tests, if related to clang, are placed in clang/test/, usually CodeGen or CodeGenCXX. This means that sometimes we have to duplicate tests. If useful, you can add a comment to tell future readers how to find related tests.

earnol replied:
Duplication looks really ugly. Is there is a way to run IR test and runtime test on the same code?

earnol replied:
Anyways, the test was duplicated.

earnol commented:
Ping!
Any additional suggestions for the improvements?
Selfishly want to move it forward :)

earnol commented:
One more ping as week passed from the previous one.

AaronBallman reviewed (commented):
The changes seem reasonable to me, but I don't have a lot of expertise with compiler-rt. Should the changes come with a release note though?

ilovepi commented on compiler-rt/test/ubsan/TestCases/Integer/bit-int-pass.c:
```diff
@@ -0,0 +1,42 @@
+  return (r > 0) ? r : 0;
+}
+
+// In those examples the file is expected to compile&run with no diagnostics
```
```suggestion
// In those examples the file is expected to compile & run with no diagnostics
```

earnol replied:
What typo you are talking about? 
'compile & run' vs 'compile&run'?

ilovepi replied:
yes.

earnol replied:
It did not look like a typo for me because i always used '&' without surrounding spaces. So it was "intentional" typo. Unfortunately for me, googling showed  'compile & run' is more widespread compared to 'compile&run'.
Thus ==> accepted.

ilovepi reviewed (approved):
Aside from a typo, this is LGTM from me. I'd give other reviewers a chance to respond before landing. I know this has been sitting for some time w/o feedback, so I'd hazard that they're fine with it, but I'd still give them a bit of time to check once more.

MaskRay commented on compiler-rt/test/ubsan/TestCases/Integer/bit-int.c:
```diff
@@ -0,0 +1,170 @@
+// RUN: %clang -Wno-constant-conversion -Wno-array-bounds -Wno-division-by-zero -Wno-shift-negative-value -Wno-shift-count-negative -Wno-int-to-pointer-cast -O0 -fsanitize=alignment,array-bounds,bool,float-cast-overflow,implicit-integer-sign-change,implicit-signed-integer-truncation,implicit-unsigned-integer-truncation,integer-divide-by-zero,nonnull-attribute,null,nullability-arg,nullability-assi ...
```
The -fsanitize= option lists a lot of unused checks. Clean up them to list only used checks

MaskRay commented on compiler-rt/test/ubsan/TestCases/Integer/bit-int-pass.c:
```diff
@@ -0,0 +1,42 @@
+// RUN: %clang -Wno-constant-conversion -Wno-array-bounds -Wno-division-by-zero -Wno-shift-negative-value -Wno-shift-count-negative -Wno-int-to-pointer-cast -O0 -fsanitize=alignment,array-bounds,bool,float-cast-overflow,implicit-integer-sign-change,implicit-signed-integer-truncation,implicit-unsigned-integer-truncation,integer-divide-by-zero,nonnull-attribute,null,nullability-arg,nullability-assi ...
```
You can merge these tests into `bit-int.c` by utilizing `--implicit-check-not="runtime error:"`

MaskRay commented on compiler-rt/lib/ubsan/ubsan_value.h:
```diff
@@ -125,6 +137,26 @@ class TypeDescriptor {
+    CHECK(isSignedBitIntTy());
+    // Scan Name for zero and return the next address
+    const char *p = getTypeName();
+    while (*p != '\0') {
```
drop braces https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

MaskRay commented on compiler-rt/lib/ubsan/ubsan_value.h:
```diff
@@ -125,6 +137,26 @@ class TypeDescriptor {
   }
 
+  const char *getBitIntBitCountPointer() const {
+    CHECK(isBitIntTy());
```
CHECK unconditionally compiles to additional code. perhaps DCHECK

MaskRay commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3298,22 +3315,40 @@ LValue CodeGenFunction::EmitPredefinedLValue(const PredefinedExpr *E) {
+    // making it <diagnostic-like type name>.'\0'.<32-bit width>.
+    if (T->isSignedIntegerType() && T->getAs<BitIntType>()) {
+      // Do a sanity checks as we are using 32-bit type to store bit length.
+      assert((getContext().getTypeSize(T) > 0) &&
```
The prevailing style uses `A > B && `, not `(A > B) && `

another choice is `assert(("xxxx", A > B))`

MaskRay commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3324,6 +3359,20 @@ llvm::Constant *CodeGenFunction::EmitCheckTypeDescriptor(QualType T) {
+  if (IsBitInt) {
+    // The Structure is: 0 to end the string, 32 bit unsigned integer in target
+    // endianness, zero.
+    char S[6] = {'\0', '\0', '\0', '\0', '\0', '\0'};
```
How is the trailing zero byte used?

AdamMagierFOSS replied:
Probably a bit too late to the party and I struggle with following discussion history on GitHub, but what was the reasoning behind storing the bit width encoded in the name of the type versus parsing out the bit width from the name of the type itself in the runtime (e.g. parsing out the value 37 from '_BitInt(37)' instead of extracting the literal 37 from '_BitInt(37)\x00\x25\x00\x00\x00' )? I would imagine that the former would be easier since there wouldn't be any platform dependence when traversing the string, and I don't think it would incur too much of an extra performance hit (the solution might look a bit kludgey though)

MaskRay commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3298,22 +3315,40 @@ LValue CodeGenFunction::EmitPredefinedLValue(const PredefinedExpr *E) {
+    TypeKind = TK_Integer;
     TypeInfo = (llvm::Log2_32(getContext().getTypeSize(T)) << 1) |
                (T->isSignedIntegerType() ? 1 : 0);
+    // Follow suggestion from https://github.com/llvm/llvm-project/issues/64100
```
Consider removing this link. Authority resides in code or formal documentation, not an /issues/ link.

Commit:
[ubsan] Display correct runtime messages for negative _BitInt

Without this patch compiler-rt ubsan library has a bug displaying
incorrect values for variables of the _BitInt (previousely called
_ExtInt) type. This patch affects both: generation of metadata
inside code generator and runtime part. The runtime part provided only
for i386 and x86_64 runtimes. Other runtimes should be updated to take
full benefit of this patch.
The patch is constructed the way to be backward compatible and
int and float type runtime diagnostics should be unaffected for not
yet updated runtimes.

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>
Co-authored-by: Paul Kirth <paulkirth@google.com>

AaronBallman commented on clang/test/CodeGen/bit-int-ubsan.c:
```diff
@@ -0,0 +1,96 @@
+
+// The runtime test checking the _BitInt ubsan feature is located in compiler-rt/test/ubsan/TestCases/Integer/bit-int.c
+
+#include <stdint.h>
+#include <stdio.h>
```
Please replace these includes with the declarations needed from them -- this makes the tests more stable because they won't be relying on whatever system CRT headers are installed.

thurstond commented:
One of the tests is failing:
```
compiler-rt/test/ubsan/TestCases/Integer/bit-int-pass.c:31:15: error: CHECK-NOT: excluded string found in input
// CHECK-NOT: runtime error:
              ^
<stdin>:1:110: note: found here
compiler-rt/test/ubsan/TestCases/Integer/bit-int-pass.c:11:10: runtime error: implicit conversion from type '_BitInt(37)' of value 60129542145 (64-bit, signed) to type 'uint32_t' (aka 'unsigned int') changed the value to 1 (32-bit, unsigned)
                                                                                                             ^~~~~~~~~~~~~~
```

earnol commented:
On what platform this failure is observed?

earnol commented:
> One of the tests is failing:
> 
> ```
> $ LIT_FILTER=bit-int-pass ninja check-ubsan
> 
> compiler-rt/test/ubsan/TestCases/Integer/bit-int-pass.c:31:15: error: CHECK-NOT: excluded string found in input
> // CHECK-NOT: runtime error:
>               ^
> <stdin>:1:110: note: found here
> compiler-rt/test/ubsan/TestCases/Integer/bit-int-pass.c:11:10: runtime error: implicit conversion from type '_BitInt(37)' of value 60129542145 (64-bit, signed) to type 'uint32_t' (aka 'unsigned int') changed the value to 1 (32-bit, unsigned)
>                                                                                                              ^~~~~~~~~~~~~~
> ```

On what target you observe the failure?

thurstond commented:
> > One of the tests is failing:
> > ```
> > $ LIT_FILTER=bit-int-pass ninja check-ubsan
> > 
> > compiler-rt/test/ubsan/TestCases/Integer/bit-int-pass.c:31:15: error: CHECK-NOT: excluded string found in input
> > // CHECK-NOT: runtime error:
> >               ^
> > <stdin>:1:110: note: found here
> > compiler-rt/test/ubsan/TestCases/Integer/bit-int-pass.c:11:10: runtime error: implicit conversion from type '_BitInt(37)' of value 60129542145 (64-bit, signed) to type 'uint32_t' (aka 'unsigned int') changed the value to 1 (32-bit, unsigned)
> >                                                                                                              ^~~~~~~~~~~~~~
> > ```
> 
> On what target you observe the failure?

x86-64 Linux

rorth commented:
The patch also broke the [Solaris/sparcv9 buildbot](https://lab.llvm.org/buildbot/#/builders/13/builds/1579).

earnol commented:
Added https://github.com/llvm/llvm-project/pull/104494 to address failures with non-x86 targets as those platforms might produce failures.

nikic commented:
The clang test also fails on s390x (big endian):
```
RUN: at line 2: /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/llvm/s390x-redhat-linux-gnu/bin/clang -cc1 -internal-isystem /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/llvm/s390x-redhat-linux-gnu/bin/../lib/clang/20/include -nostdsysteminc -Wno-constant-conversion -Wno-array-bounds -Wno-division-by-zero -Wno-shift-negative-value -Wno-shift-count-negative -Wno-int-to-pointer-cast -fsanitize=array-bounds,enum,float-cast-overflow,integer-divide-by-zero,implicit-unsigned-integer-truncation,implicit-signed-integer-truncation,implicit-integer-sign-change,unsigned-integer-overflow,signed-integer-overflow,shift-base,shift-exponent -O0 -emit-llvm -o - /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c | /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/llvm/s390x-redhat-linux-gnu/bin/FileCheck /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c
+ /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/llvm/s390x-redhat-linux-gnu/bin/clang -cc1 -internal-isystem /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/llvm/s390x-redhat-linux-gnu/bin/../lib/clang/20/include -nostdsysteminc -Wno-constant-conversion -Wno-array-bounds -Wno-division-by-zero -Wno-shift-negative-value -Wno-shift-count-negative -Wno-int-to-pointer-cast -fsanitize=array-bounds,enum,float-cast-overflow,integer-divide-by-zero,implicit-unsigned-integer-truncation,implicit-signed-integer-truncation,implicit-integer-sign-change,unsigned-integer-overflow,signed-integer-overflow,shift-base,shift-exponent -O0 -emit-llvm -o - /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c
+ /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/llvm/s390x-redhat-linux-gnu/bin/FileCheck /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c
/builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c:11:12: error: CHECK: expected string not found in input
 // CHECK: constant { i16, i16, [20 x i8] } { i16 2, i16 13, [20 x i8] c"'_BitInt(37)'\00%\00\00\00\00\00" }
           ^
<stdin>:7:100: note: scanning from here
@0 = private unnamed_addr constant { i16, i16, [8 x i8] } { i16 1, i16 32, [8 x i8] c"'float'\00" }
                                                                                                   ^
<stdin>:8:27: note: possible intended match here
@1 = private unnamed_addr constant { i16, i16, [20 x i8] } { i16 2, i16 13, [20 x i8] c"'_BitInt(37)'\00\00\00\00%\00\00" }
                          ^

Input file: <stdin>
Check file: /builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = '/builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c' 
            2: source_filename = "/builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c" 
            3: target datalayout = "E-m:e-i1:8:16-i8:8:16-i64:64-f128:64-v128:64-a:8:16-n32:64" 
            4: target triple = "s390x-redhat-linux-gnu" 
            5:  
            6: @.src = private unnamed_addr constant [111 x i8] c"/builddir/build/BUILD/llvm-project-ec29660c44e5e73d3b78f4884f9178036563fb25/clang/test/CodeGen/bit-int-ubsan.c\00", align 2 
            7: @0 = private unnamed_addr constant { i16, i16, [8 x i8] } { i16 1, i16 32, [8 x i8] c"'float'\00" } 
check:11'0                                                                                                        X error: no match found
            8: @1 = private unnamed_addr constant { i16, i16, [20 x i8] } { i16 2, i16 13, [20 x i8] c"'_BitInt(37)'\00\00\00\00%\00\00" } 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:11'1                               ?                                                                                                  possible intended match
            9: @2 = private unnamed_addr global { { ptr, i32, i32 }, ptr, ptr } { { ptr, i32, i32 } { ptr @.src, i32 10, i32 19 }, ptr @0, ptr @1 } 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10: @3 = private unnamed_addr constant { i16, i16, [32 x i8] } { i16 0, i16 10, [32 x i8] c"'uint32_t' (aka 'unsigned int')\00" } 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           11: @4 = private unnamed_addr global { { ptr, i32, i32 }, ptr, ptr, i8, i32 } { { ptr, i32, i32 } { ptr @.src, i32 12, i32 10 }, ptr @1, ptr @3, i8 2, i32 0 } 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           12: @5 = private unnamed_addr constant { i16, i16, [6 x i8] } { i16 0, i16 11, [6 x i8] c"'int'\00" } 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           13: @6 = private unnamed_addr global { { ptr, i32, i32 }, ptr } { { ptr, i32, i32 } { ptr @.src, i32 16, i32 21 }, ptr @5 } 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
>>>>>>
```

Please pass an explicit `-triple` to `%clang_cc1` if the test is not supposed to be big endian compatible.

AaronBallman commented:
@earnol -- given the amount of breakage, can you revert the changes to get all bots back to green, then we can re-land once you have everything fixed up?

earnol commented:
How i can see the full amount of breakage on different platforms proactively?
I have created a PR: https://github.com/llvm/llvm-project/pull/104607 to address the big endian issue with the test as my // REQUIRES: x86-registered-target turned out not to be enough to exclude running test in the big endian platforms.

earnol commented:
Since automatic revert is not working i have applied https://github.com/llvm/llvm-project/pull/104607 and monitoring builders.

earnol commented:
> Please pass an explicit `-triple` to `%clang_cc1` if the test is not supposed to be big endian compatible.

This had been done in https://github.com/llvm/llvm-project/pull/104607. Please check.

AaronBallman commented:
> How i can see the full amount of breakage on different platforms proactively? I have created a PR: #104607 to address the big endian issue with the test as my // REQUIRES: x86-registered-target turned out not to be enough to exclude running test in the big endian platforms. If you insisting on reverting PR completely it is also fine, but it would be useful to see the full list of errors on different platforms so i would be able to address those.

Ah, if that's resolving the issues, then that's totally fine to fix forward rather than revert. Thank you!

vitalybuka commented:
> > How i can see the full amount of breakage on different platforms proactively? I have created a PR: #104607 to address the big endian issue with the test as my // REQUIRES: x86-registered-target turned out not to be enough to exclude running test in the big endian platforms. If you insisting on reverting PR completely it is also fine, but it would be useful to see the full list of errors on different platforms so i would be able to address those.
> 
> Ah, if that's resolving the issues, then that's totally fine to fix forward rather than revert. Thank you!

Also it's totally fine to revert/reland multiple times, incrementally detecting and fixing breakages related to the patch.
Broken build bot will more likely hide other breakages, making it harder to restore bots into green state.

earnol commented:
As i see the fedora build is fine: https://github.com/fedora-llvm-team/llvm-snapshots/issues/649.

bjope commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3298,22 +3315,40 @@ LValue CodeGenFunction::EmitPredefinedLValue(const PredefinedExpr *E) {
+    // Follow suggestion from discussion of issue 64100.
+    // So we can write the exact amount of bits in TypeName after '\0'
+    // making it <diagnostic-like type name>.'\0'.<32-bit width>.
+    if (T->isSignedIntegerType() && T->getAs<BitIntType>()) {
```
Is this whole patch limited to fixing the non-power-of-2 problem for signed BitInt? I couldn't see any test involving unsigned BitInt, and the check for signed here makes me think that we still got the same bug for type descriptors for unsigned BitInt.

If that is true, was there any specific reason for not fixing the type descriptors for unsigned BitInt as well directly?
If for example the reasoning is that the runtimes currently do not try to print the size anyway for any failures involving unsigned BitInt, maybe it had been worth mentioning that here. Because now it just looks broken to me that we still can't provide a correct type descriptor for unsigned BitInt. Right?

earnol replied:
The whole patch is limited to signed non-power-of-2 _BitInt only. Unsigned BitInts are not affected by the patch and did not have an issue to begin with. Please see the issue: https://github.com/llvm/llvm-project/issues/64100.

jakubjelinek commented on compiler-rt/lib/ubsan/ubsan_value.h:
```diff
@@ -125,6 +137,25 @@ class TypeDescriptor {
+  unsigned getIntegerBitCount() const {
+    DCHECK(isIntegerTy());
+    if (isSignedBitIntTy())
+      return *reinterpret_cast<const u32 *>(getBitIntBitCountPointer());
```
This will not work on strict alignment targets.  I'd suggest instead of trying to reinterpret the bits memcpy them into u32 and return that, the compilers will optimize that properly.

earnol replied:
Well. I do agree with this statement, however do we have a way to know whether target has a strict alignment or not? Preferably on the define level or you would suggest use always use memcpy? 
I considered this option, but was really reluctant to add memcpy dependency into libubsan.
So may be the better option is gather it byte by byte a strict alignment platform.

bjope replied:
For example Value::getFloatValue() in ubsan_value.cpp is using internal_memcpy from sanitizer_libc. Maybe that is an option here.

jakubjelinek commented on compiler-rt/lib/ubsan/ubsan_value.h:
```diff
@@ -103,6 +103,13 @@ class TypeDescriptor {
     /// integer type.
     TK_Float = 0x0001,
+    /// An _BitInt(N) type. Lowest bit is 1 for a signed value, 0 for an
+    /// unsigned value. Remaining bits are log_2(bit_width). The value
```
bit_width of what?  The main problem is that the _BitInt ABI is only defined in the x86-64 and aarch64 psABIs, for i686 the word was that the ABI should be whatever GCC implemented, but nobody updated the ia32 psABI yet.
I think ARM 32-bit is defined too, some other targets are working on it.
As documented in LLVM, it implements _BitInt everywhere but notes that the ABI can change (GCC took a different approach, only implementing it for arches which agree on an ABI).
So, if compiler-rt implements _BitInt support, it better be able to encode the various ABIs which will be used for _BitInt on various arches.
The current ABIs usually have some special cases for smaller number of bits, but those generally can be encoded as normal integer (sign or zero extended say to long long or __int128).
For the large _BitInts, x86_64 uses little endian ordered array of 64-bit limbs, aarch64 little or big endian ordered array of 128-bit limbs, ia32 little endian ordered array of 32-bit limbs.
But I was told the powerpc folks are considering little endian ordering of limbs on both ppc64le and ppc64 big-endian.
So, I think you certainly want to encode the limb size and the N from _BitInt(N) and whether it is signed or unsigned and possibly (e.g. for ppc64) also whether the limbs are ordered from least significant to most or vice versa (bit ordering within limb would be normally dependent on endianity).
So, I think that bit_width in the bits above sign should be the limb size, not _BitInt size rounded up or whatever you're implementing right now.  And encode somewhere the endianity too.

earnol replied:
It's interesting problem. I have not changed the way _BitInts are encoded in the memory. My understanding is that might not be right approach to have too many meanings embedded into one field depending on the context. Please note we have only 15 bits here (lowest bit is a sign). Standard requires us to support N up to BITINT_MAXWIDTH which is quite high.
The standard says: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2709.pdf
_The macro BITINT_MAXWIDTH represents the maximum width N supported in the declaration
of a bit-precise integer type (6.2.5) in the type specifier _BitInt(N). The value
BITINT_MAXWIDTH shall expand to a value that is greater than or equal to the value of
ULLONG_WIDTH._

It means it will not fit :(
Also why we do need endianess? Cannot we always use endianess of the target platform? We always know it at the moment of code generation.

bjope replied:
The new thing in this patch is that the size of the _BitInt type being described is stored after the type string. It can be read using `getIntegerBitCount()`.

Thus, to get the size of the type being described one should no longer use  `getIntegerBitWidth()` but instead `getIntegerBitCount()`. The latter would give correct result for both TK_Integer and TK_BitInt, while the former doesn't work for BitInt.  IMHO it had been better to keep the name `getIntegerBitWidth()` for this purpose. Now I figure most/all old uses of `getIntegerBitWidth()` actually should be replaced with calls to `getIntegerBitCount()`, but that has not happened as far as I can tell.

The new behavior of  `getIntegerBitWidth()` is merely to indicate the size used to store the Value being represented when it does not fit inside the ValueHandle. So there shouldn't really be any need to use that function outside `Value::getSIntValue()` (assuming that `getIntegerBitCount()` would know how to get the size of TK_Integer via the TypeInfo field by itself).

And unfortunatly there isn't much information about how those non-inlined values are store. The `Value::getSIntValue()` functions is just reading an s64 or s128 from memory (and that is what it traditionally has been doing for TK_Integer). But then I think the IR emitted by clang should cast the value to i64/i128 instead of doing  
```
    RawAddress Ptr = CreateDefaultAlignTempAlloca(V->getType());
    Builder.CreateStore(V, Ptr);
```
Now we for example get things like `store i66 %x, ptr %tmp` which is storing a non byte-sized value to memory, which then is loaded by the runtime as a 128-bit value.
Worth noting is that after commit 9ad72df55cb74b2919327 (#91364) the frontend emits byte-sized loads/stores for _BitInt. So I really think the store emitted by CodeGenFunction::EmitCheckValue for the BitInt values should zero/sign extended similarly (as if they were casted to "s64" or "s128" before being stored). That would at least make the code a bit more consistent (even though the question about supporting values larger than 128-bits would remain).

jakubjelinek commented:
I'm not suggesting to encode the number of limbs anywhere, I'm suggesting encoding the bit precision of a limb somewhere.  And the limb ordering.
On little endian of bits in a limb and little endian ordering of limbs in the limb array, at least if the limbs are sane (have precision multiple of char precision and there are no padding bits in between), the actual limb precision might seem to be irrelevant, all you care about is the N from {,{un,}signed }_BitInt(N) and whether it is unsigned or signed, so you can treat the passed pointer say as an array of 8-bit limbs, N / 8 limbs with full 8 bits and if N % 8, the last limb containing some further bits (in some ABIs that will be required to be sign or zero extended, in other ABIs the padding bits will be undefined, but on the libubsan side you can always treat them as undefined and always manually extend).
Or treat it as 16-bit limbs, or 32-bit limbs, or 64-bit limbs, or 128-bit limbs, for the higher perhaps with doing the limb reads again using internal_memcpy so that you don't impose some alignment requirement perhaps the target doesn't have.
But on big-endian, I think knowing the limb precision/size is already essential (sure, just a theory for now, GCC right only only supports _BitInt on little-endian targets because those are the only ones that have specified their ABI).
E.g. I believe _BitInt(513) big-endian with big-endian limb ordering would be for 32-bit limbs 17 limbs, the first one containing just one bit (the most significant of the whole number) and the remaining ones each 32 bits, while for 64-bit limbs 9 limbs, the first one containing just one bit and the remaining ones 64 bits each; and for 128-bit limbs 5 limbs, the first one just one bit, the remaining 128 bits each.  You can't decode these without knowing the limb size, the data looks different in memory.
And then there is the possibility of big-endian limbs with little-endian ordering of the limbs in the array.
As the 15 bits of the current precision used e.g. for normal integers is clearly insufficient  to express supported BITINT_MAXWIDTH (8388608 in clang, 65535 right now in GCC), my suggestion is to use another bit for the limb ordering
(say 0 little endian, 1 big endian) and the reaming 14 bits for the limb precision (whether log2 or not doesn't matter that much).
As for the actual _BitInt precision after the type name, one option is what you currently implemented, i.e. always use 32-bit integer in memory there, plus the extra '\0' termination if you really think it is needed, IMHO it is just a waste, and another
option is to use say uleb128 encoding of it.
